cirq-core/cirq/ops/boolean_hamiltonian.py

```diff
@@ -0,0 +1,372 @@
+    """Compares two Gray-encoded binary numbers.
+
+    Args:
+        k1: A tuple of ints, representing the bits that are one. For example, 6 would be (1, 2).
+        k2: The second number, represented similarly as k1.
+        flip: Whether to flip the comparison.
+
+    Returns:
+        -1 if k1 < k2 (or +1 if flip is true)
+        0 if k1 == k2
+        +1 if k1 > k2 (or -1 if flip is true)
+    """
+    max_1 = k1[-1] if k1 else -1
+    max_2 = k2[-1] if k2 else -1
+    if max_1 != max_2:
+        return -1 if (max_1 < max_2) ^ flip else 1
+    if max_1 == -1:
+        return 0
+    return _gray_code_comparator(k1[0:-1], k2[0:-1], not flip)
+
+
+def _simplify_cnots_pairs(
+    cnots: List[Tuple[int, int]], flip_control_and_target: bool
+) -> Tuple[bool, List[Tuple[int, int]]]:
+    """Simplifies CNOT pairs according to equations 9 and 10 of [4].Simplifies
+
+    CNOT(i, j) @ CNOT(k, j) = CNOT(k, j) @ CNOT(i, j)
+    ───@───────       ───────@───
+       │                     │
+    ───X───X───   =   ───X───X───
+           │             │
+    ───────@───       ───@───────
+
+    Args:
+        cnots: A list of CNOTS, encoded as integer tuples (control, target).
+        flip_control_and_target: Whether to flip control and target.
+
+    Returns:
+        A Boolean that tells whether a simplification has been performed.
+        The CNOT list, potentially simplified.
+    """
+    x, y = (0, 1) if flip_control_and_target else (1, 0)
+
+    i = 0
+    qubit_to_index: Dict[int, int] = {}
+    for j in range(1, len(cnots)):
+        if cnots[i][x] != cnots[j][x]:
+            # The targets (resp. control) don't match, so we reset the search.
+            i = j
+            qubit_to_index = {cnots[j][y]: j}
+            continue
+
+        if cnots[j][y] in qubit_to_index:
+            k = qubit_to_index[cnots[j][y]]
+            # The controls (resp. targets) are the same, so we can simplify away.
+            cnots = [cnots[n] for n in range(len(cnots)) if n != j and n != k]
+            return True, cnots
+
+        qubit_to_index[cnots[j][y]] = j
+
+    return False, cnots
+
+
+def _simplify_cnots_tripplets(
+    cnots: List[Tuple[int, int]], flip_control_and_target: bool
+) -> Tuple[bool, List[Tuple[int, int]]]:
+    """Simplifies CNOT pairs according to equations 11 of [4].
+
+    CNOT(i, j) @ CNOT(j, k) == CNOT(j, k) @ CNOT(i, k) @ CNOT(i, j)
+    ───@───────       ───────@───@───
+       │                     │   │
+    ───X───@───   =   ───@───┼───X───
+           │             │   │
+    ───────X───       ───X───X───────
+
+    Args:
+        cnots: A list of CNOTS, encoded as integer tuples (control, target).
+        flip_control_and_target: Whether to flip control and target.
+
+    Returns:
+        A Boolean that tells whether a simplification has been performed.
+        The CNOT list, potentially simplified.
+    """
+    x, y = (0, 1) if flip_control_and_target else (1, 0)
```

Comment:
```suggestion
    target, control = (0, 1) if flip_control_and_target else (1, 0)
```
pick variable names that describe their responsibility.